### ISSUE TITLE
Vastly different results when running the tensorflow or tensorrt models

### ISSUE THREAD

**Issue by ghylander** [NVIDIA/TensorRT#1175]
## Description

I have a trained tensorflow model, with a certain reported valdiation acurracy and confusion amtrix

after deploying it on a jsetson nano (tensorflow -> onnx (tf2onnx), onnx -> tensorrt), the model was performing quite bad
after copying the valdiation dataset from the training machine to the jetson and running a confusion matrix script, the results are VERY different

this is an image classification problem

## Environment

**TensorRT Version**: 7.1.3
**NVIDIA GPU**: jetson nano
**Jetpack version**: JetPack 4.4.1
**Operating System**: ubuntu 18.04
**Python Version (if applicable)**: 3.8.5
**Tensorflow Version (if applicable)**: 2.5 dev

## Relevant Files
I had several models to test how each would fare in real world usage. This is one of them. Top matrix is the tensorflow training result, bottom one is running in the jetson nano.

![WhatsApp Image 2021-04-08 at 09 30 39](https://user-images.githubusercontent.com/74593034/113991417-b4068b00-9852-11eb-8c31-513dcab35f97.jpeg)

Same as above but a different even worse performing model
![WhatsApp Image 2021-04-08 at 09 43 02](https://user-images.githubusercontent.com/74593034/113991399-af41d700-9852-11eb-8985-ecdb43e7d5e1.jpeg)

Both the training machine and jetson have the exact same validation dataset these confusion matrices are elaboratedwith
## Steps To Reproduce

export from tensorflow to onnx using tf2onnx
export onnx to tenorrt engine

**Comment by ttyio** [comment]
Hello @ghylander , could you provide onnx file to us for debugging? thanks

**Comment by ghylander** [comment]
hello, 
After a lot more trial, it turns out that the conversion to either TRT or onnx produces incorrect models, seemingly at random. After exporting a batch of 10 models from tensorflow to onnx and later to trt, only 2 of them had correct behaviour, while the rest all performed like the 2nd image.

Nevertheless, the issue in the first image still stands.
We discovered this when applying conv net visualization techniques to wrongly classified images (false negatives and false positives).
We were running an image through trt, and if it was a false negative or a false positive, we would take the image and run it through tensorflow to access the gradients and get the heatmaps (adapted from chollet's book). 

The problem that lead us to find this issue, which still persists, is that wrongly classified images on the jetson,  when run through tensorflow produce the opposite result

I understand the process of creating a TRT goes through simplifying, merging, etc the layers of the base model, and i expected some difference. However, most of the time,when there is a difference in classification, it's a case of [0.99, 0,01] on the jetson and [0.01, 0,99] using tensorflow (and viceversa).

As an example, when running one particular image through tensorflow, the output is the following: 

![image](https://user-images.githubusercontent.com/74593034/114824402-1caab600-9dc5-11eb-9397-6daeaaefcba5.png)

in the first output block, you can see a verification we run to check the array entering both networks is the same. The dimensions first, then a vector with the mean of each channel of the image. Lastly, the second output block shows the predictions vector, showing the input image was classsified as class 0

Running inference of the same input image on the jetson produces the following result:

![image](https://user-images.githubusercontent.com/74593034/114824353-0dc40380-9dc5-11eb-8e1e-247a9d36b84e.png)

You can see the same information here. Dimensions, mean of each channel, and predictions vector. The predictions vector however now shows a clear class 1 classificaton for the same input image

**Comment by ghylander** [comment]
as we continue to look into this issue, i ran some more tests

i took a test dataset images, and fed them to the network using tensorflow and TRT. I recorded the classification for each image in a CSV file. I then compared the CSV against each other. Taking the classification made by tensorflow as the 'correct' or 'expected' output of the model, here is the result:
![image](https://user-images.githubusercontent.com/74593034/115519919-ecae5780-a289-11eb-8653-495bbff431bd.png)

Out of over 1000 images, over 50% of the images were differently classified on the jetson as on Tensorflow

In addition to this, either the exporting to onnx or TRT format seems to be unreliable. During the test i've been running, i've had to export models from tensorflwo to onnx, and later to TRT. The resulting .engine files had different performance levels, with some of them outright classifying all fed images as either of the classes (similarly to the second image in my original post)

**Comment by ttyio** [comment]
Hello @ghylander , for the question "either the exporting to onnx or TRT format seems to be unreliable."
 We have an opensource tool to compare between trt and onnxruntime, 
https://github.com/NVIDIA/TensorRT/tree/master/tools/Polygraphy

the command line looks like this:

      polygraphy run --trt --onnxrt your.onnx

Could you take a try? thanks!

**Comment by ghylander** [comment]
If i understood the documentation correctly, this comapres the execution of the model when run by tensorrt and onnx

using this information, i can infere that running the command both for a onnx model and trt model should produce the same results (for each runtime), hence verifying that the conversion from onnx to trt was succesful with no errors

**Comment by ttyio** [comment]
Yes @ghylander , and due to floating point arithmetic sequence, there will be slight difference, we have `--rtol` `--atol` to set the tolerance in polygraphy.

**Comment by ttyio** [comment]
Close since no activity for more than 3 weeks, please reopen if you still have question, thanks!
